File: java-client/src/main/java/io/fluxcapacitor/javaclient/tracking/Tracking.java
/*
 * Copyright (c) Flux Capacitor IP B.V. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.fluxcapacitor.javaclient.tracking;

import io.fluxcapacitor.common.Registration;
import io.fluxcapacitor.javaclient.FluxCapacitor;

import java.util.Arrays;
import java.util.List;

/**
 * The Tracking interface provides a mechanism to start and manage the tracking of messages by a given set of handlers.
 */
public interface Tracking extends AutoCloseable {

    /**
     * Starts the tracking process using the specified FluxCapacitor instance and the provided handlers.
     *
     * @param fluxCapacitor the FluxCapacitor instance to be used for tracking
     * @param handlers      the handlers responsible for processing tracked messages
     * @return a Registration instance that can be used to manage and cancel the tracking process
     */
    default Registration start(FluxCapacitor fluxCapacitor, Object... handlers) {
        return start(fluxCapacitor, Arrays.asList(handlers));
    }

    /**
     * Starts tracking messages using the provided FluxCapacitor and a list of handlers.
     *
     * @param fluxCapacitor the FluxCapacitor used to manage the message tracking process
     * @param handlers      a list of handlers that process the tracked messages
     * @return a Registration instance that can be used to cancel the tracking process
     */
    Registration start(FluxCapacitor fluxCapacitor, List<?> handlers);

    /**
     * Closes the tracking process, releasing any resources held by it. This method is invoked automatically when
     * {@link FluxCapacitor} is shut down.
     */
    @Override
    void close();
}

File: js-client/src/Tracking.ts
/*
 * Copyright (c) 2016-2017 Flux Capacitor.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Append, Message, MessageBatch, Read, ReadResult, StorePosition} from './Api';
import {WebSocketService} from './WebSocketService';

export interface ConsumerService {
    read(consumer : string, maxSize : number, maxTimeout : number): Promise<MessageBatch>;

    storePosition(consumer : string, segment : [number, number], lastIndex : any) : void;
}

export interface ProducerService {
    publish(messages : Message[]) : void;
}

export class WebSocketConsumerService extends WebSocketService implements ConsumerService {
    read(consumer: string, maxSize: number, maxTimeout: number): Promise<MessageBatch> {
        const result : Promise<ReadResult> = this.sendRequest(new Read(consumer, maxSize, maxTimeout, 0));
        return result.then(success => MessageBatch.deserialize(success.messageBatch));
    }

    storePosition(consumer: string, segment: [number, number], lastIndex: any) : void {
        this.send(new StorePosition(consumer, segment, lastIndex));
    }
}

export class WebSocketProducerService extends WebSocketService implements ProducerService {
    publish(messages: Message[]) : void {
        this.send(new Append(messages));
    }
}

export async function startTracking(consumerName: string, consumer: Function, service: ConsumerService) {
    let stopped = false;
    while (!stopped) {
        const batch = await service.read(consumerName, 64, 60000);
        batch.messages.map(msg => {
            return {
                type: msg.type(),
                payload: msg.payload(),
                revision: msg.revision()
            }
        }).forEach(msg => consumer(msg));
        if (batch.lastIndex) {
            service.storePosition(consumerName, batch.segment, batch.lastIndex);
        }
    }
}